Définition de l’architecture du réseau de neurones

In [1]:
import numpy as np
from keras import models, layers
from keras.datasets import boston_housing
import matplotlib.pyplot as plt
from keras.models import Sequential 
from keras.layers import Activation, Dense 
from keras import initializers 
from keras import regularizers 
from keras import constraints 

In [2]:
def def_architecture():
    m = models.Sequential()
    m.add(Dense(64, input_shape=(13, ), activation='relu'))
    m.compile(optimizer='rmsprop',loss='mse')
    return m

Question 4)

In [ ]:
#Réalisé à l'aide du pseudo-code, je le laisse même si je ne crois pas que c'est ce qui était demandé
# import random
#def perceptron_une_couche_64(DE,De):
#    W = np.random.randint(-100, 100, size=(14,65))
#    W = W*0.01
#    lr = 0.01
#    OS = 0
#    t = 1
#    while (OS-t)**2 > 0.1 :
#        L = 0
#        while np.size(De) != 0 :
#            ir = random.randint(0,np.size(De)-1)
#            o = np.zeros(14)
#            o[0] = 1
#            for i in range(13):
#                o[i+1] = DE[ir,i]
#            t = De[ir-1]
#            O = np.zeros(65)
#            O[0] = 1
#            for j in range(1,65):
#                for k in range(14):
#                    O[j] = O[j] + o[k]*W[k,j]
#                if O[j] < 0:
#                    O[j] = 0
#                else :
#                    O[j] = O[j]
#            OS = np.sum(O)
#            L = L + (1/2)*((OS-t)**2)
#            DELTAS = OS-t
#            DELTA = np.zeros(65)
#            for j in range(65):
#                if OS >= 0:
#                    DELTA[j] = DELTAS
#                else:
#                    DELTA[j] = 0
#            delta = np.zeros(14)
#            for i in range(14):
#                for k in range(65):
#                    if O[j] >= 0:
#                        delta[i] = delta[i] + DELTA[k]*W[i,k]
#                    else:
#                        delta[i] = 0
#            for i in range(14):
#                for j in range(65):
#                    W[i,j] = W[i,j] + lr*DELTA[j]*o[i]
#            np.delete(De, ir)
#            np.delete(DE, ir)
#    return OS

Question 5)

Sur la première couche on a un paramètre pour chaque connexion neuronale aux neurones de la couche cachée, c'est le poids de chaque connexion neuronale. Comme on a 13 entrées + 1 neurone de biais, on a 14*64 = 896 paramètres sur la première couche.

Sur la couche cachée on a deux paramètres pour chaque connexion neuronale impliquant la couche de sortie. On oublie pas le neurone de biais de la couche cachée, et ainsi on a 2*(64+1) = 2*65 = 130 paramètres pour la couche cachée.

Sur la couche de sortie on seulement 1 paramètre, c'est la valeur cible des estimations.

Au total le réseau neuronal est à 896 + 130 + 1 = 1027 paramètres.

Apprentissage du réseau avec validation. Retourne le modèle ainsi que les valeurs de la fonction de perte.

In [3]:
def Apprentissage(DA,Da):
    m = def_architecture()
    h = m.fit(DA,Da,validation_data=(DV,Dv),epochs=nb_époques,batch_size=1)
    return m , h.history

Chargement des bibliothèques et des modules

Chargement de la base de données Boston Housing Price.

In [4]:
(DA,Da),(DT,Dt) = boston_housing.load_data()

In [6]:
Da[104]

15.0

question 2) Ici, on centre et on normalise les données dans le but de les homogénéiser, ainsi chaque couple de données appartenant à (DA,Da) a la même importance relativement aux autres couples impliqués dans le calcul de la fonction de perte.

In [23]:
m = DA.mean(axis=0); s = DA.std(axis=0); DA-=m ; DA/=s; DT-=m; DT/=s

In [24]:
nb_époques=300
N=DA.shape[0]; NV=N//4

question 3) Ici, on partitionne notre ensemble d'entrainement en un ensemble de validation avec le premier quart de l'ensemble des éléments de l'ensemble d'entrainement, et en un autre ensemble composé du reste des éléments de l'ensemble d'entrainement.
Le but de la démarche est de continuer à améliorer l'apprentissage du modèle.

In [26]:
DV=DA[:NV]; Dv=Da[:NV]; DAp=DA[NV:]; Dap=Da[NV:]
modèle,h=Apprentissage(DAp,Dap)
DA_L=h['loss']; DV_L=h['val_loss']

Epoch 1/300
303/303 [==============================] - 1s 1ms/step - loss: 569.9056 - val_loss: 541.2836
Epoch 2/300
303/303 [==============================] - 0s 1ms/step - loss: 546.8922 - val_loss: 516.9024
Epoch 3/300
303/303 [==============================] - 0s 1ms/step - loss: 523.0638 - val_loss: 492.9336
Epoch 4/300
303/303 [==============================] - 0s 1ms/step - loss: 499.6652 - val_loss: 470.1991
Epoch 5/300
303/303 [==============================] - 0s 1ms/step - loss: 477.1375 - val_loss: 448.3931
Epoch 6/300
303/303 [==============================] - 0s 1ms/step - loss: 456.4127 - val_loss: 428.0652
Epoch 7/300
303/303 [==============================] - 0s 1ms/step - loss: 436.1565 - val_loss: 408.7157
Epoch 8/300
303/303 [==============================] - 0s 1ms/step - loss: 417.4586 - val_loss: 390.7630
Epoch 9/300
303/303 [==============================] - 0s 1ms/step - loss: 399.1582 - val_loss: 373.6704
Epoch 10/300
303/303 [==============================] -

303/303 [==============================] - 0s 1ms/step - loss: 85.1512 - val_loss: 93.8192
Epoch 156/300
303/303 [==============================] - 0s 1ms/step - loss: 85.0319 - val_loss: 93.6603
Epoch 157/300
303/303 [==============================] - 0s 1ms/step - loss: 84.9247 - val_loss: 93.4906
Epoch 158/300
303/303 [==============================] - 0s 1ms/step - loss: 84.8308 - val_loss: 93.3354
Epoch 159/300
303/303 [==============================] - 0s 1ms/step - loss: 84.7152 - val_loss: 93.1486
Epoch 160/300
303/303 [==============================] - 0s 1ms/step - loss: 84.6231 - val_loss: 92.9871
Epoch 161/300
303/303 [==============================] - 0s 1ms/step - loss: 84.5201 - val_loss: 92.8348
Epoch 162/300
303/303 [==============================] - 0s 1ms/step - loss: 84.4192 - val_loss: 92.6457
Epoch 163/300
303/303 [==============================] - 0s 1ms/step - loss: 84.3137 - val_loss: 92.4931
Epoch 164/300
303/303 [==============================] - 0s 1ms/step 

In [13]:
niter = [i for i in range(10,300)]
DA_L10 = DA_L[10:]
DV_L10 = DV_L[10:]
plt.plot(niter, DA_L10)
plt.xlabel('itération')
plt.ylabel('erreur d entrainement')
plt.plot(niter, DV_L10)
plt.xlabel('itération')
plt.ylabel('erreur de validation')
plt.title('Les erreurs en moyenne quadratique selon la quantite d iteration')
plt.show()

L'erreur de validation est plus petite que l'erreur d'entrainement pour les premières époques d'apprentissage, la valeur absolue de la différence entre les deux erreurs diminue jusqu'à ce que l'erreur de validation soi plus grande que l'erreur d'entrainement pour l'époque 28 d'apprentissage, suite à ça les erreurs continuent à décroître et l'erreur de validation reste plus grande que l'erreur d'entrainement.
La décroissance est de moins en moins fortes au fur et à mesure qu'on avance dans les époques d'apprentissage, et cela pour les deux erreurs.
Ainsi les erreurs sont minimales pour l'époque la plus grande, c'est à dire la 300ième.

In [27]:
listDA_L = []; listDV_L = []
for i in range(4):
    DV = DA[i*NV:(i+1)*NV]; Dv = Da[i*NV:(i+1)*NV];
    DAp = np.concatenate([DA[:i*NV],DA[(i+1)*NV:]],axis=0)
    Dap = np.concatenate([Da[:i*NV],Da[(i+1)*NV:]],axis=0)
    modèle,h = Apprentissage(DAp,Dap)
    listDA_L.append(h['loss']); listDV_L.append(h['val_loss'])

DA_L = [np.mean([y[i] for y in listDA_L]) for i in range(nb_époques)]
DV_L = [np.mean([y[i] for y in listDV_L]) for i in range(nb_époques)]

Epoch 1/300
303/303 [==============================] - 1s 1ms/step - loss: 570.8166 - val_loss: 543.5945
Epoch 2/300
303/303 [==============================] - 0s 1ms/step - loss: 548.5607 - val_loss: 519.5830
Epoch 3/300
303/303 [==============================] - 0s 1ms/step - loss: 524.8167 - val_loss: 495.8115
Epoch 4/300
303/303 [==============================] - 0s 1ms/step - loss: 501.2871 - val_loss: 472.3738
Epoch 5/300
303/303 [==============================] - 0s 1ms/step - loss: 478.0639 - val_loss: 449.7850
Epoch 6/300
303/303 [==============================] - 0s 1ms/step - loss: 456.2315 - val_loss: 428.2119
Epoch 7/300
303/303 [==============================] - 0s 1ms/step - loss: 434.9128 - val_loss: 407.3967
Epoch 8/300
303/303 [==============================] - 0s 1ms/step - loss: 414.0827 - val_loss: 387.8369
Epoch 9/300
303/303 [==============================] - 0s 1ms/step - loss: 394.7997 - val_loss: 369.4562
Epoch 10/300
303/303 [==============================] -

303/303 [==============================] - 0s 2ms/step - loss: 88.8725 - val_loss: 97.3803
Epoch 156/300
303/303 [==============================] - 0s 1ms/step - loss: 88.7825 - val_loss: 97.2379
Epoch 157/300
303/303 [==============================] - 0s 1ms/step - loss: 88.6646 - val_loss: 97.0998
Epoch 158/300
303/303 [==============================] - 0s 1ms/step - loss: 88.5882 - val_loss: 96.9791
Epoch 159/300
303/303 [==============================] - 0s 1ms/step - loss: 88.4899 - val_loss: 96.8719
Epoch 160/300
303/303 [==============================] - 0s 1ms/step - loss: 88.3948 - val_loss: 96.7286
Epoch 161/300
303/303 [==============================] - 0s 1ms/step - loss: 88.3072 - val_loss: 96.5642
Epoch 162/300
303/303 [==============================] - 0s 1ms/step - loss: 88.2211 - val_loss: 96.4123
Epoch 163/300
303/303 [==============================] - 0s 1ms/step - loss: 88.1436 - val_loss: 96.2922
Epoch 164/300
303/303 [==============================] - 0s 1ms/step 

303/303 [==============================] - 0s 1ms/step - loss: 378.1674 - val_loss: 330.4044
Epoch 12/300
303/303 [==============================] - 0s 1ms/step - loss: 364.5882 - val_loss: 316.9145
Epoch 13/300
303/303 [==============================] - 0s 1ms/step - loss: 351.8046 - val_loss: 304.3401
Epoch 14/300
303/303 [==============================] - 0s 1ms/step - loss: 340.4767 - val_loss: 292.8968
Epoch 15/300
303/303 [==============================] - 0s 1ms/step - loss: 329.6390 - val_loss: 282.1321
Epoch 16/300
303/303 [==============================] - 0s 2ms/step - loss: 319.6651 - val_loss: 272.3455
Epoch 17/300
303/303 [==============================] - 0s 1ms/step - loss: 310.4013 - val_loss: 263.0394
Epoch 18/300
303/303 [==============================] - 0s 1ms/step - loss: 301.6869 - val_loss: 254.6361
Epoch 19/300
303/303 [==============================] - 0s 1ms/step - loss: 293.8368 - val_loss: 246.8756
Epoch 20/300
303/303 [==============================] - 0s 

Epoch 166/300
303/303 [==============================] - 0s 1ms/step - loss: 91.7405 - val_loss: 57.0797
Epoch 167/300
303/303 [==============================] - 0s 1ms/step - loss: 91.5373 - val_loss: 56.9025
Epoch 168/300
303/303 [==============================] - 0s 1ms/step - loss: 91.3286 - val_loss: 56.7678
Epoch 169/300
303/303 [==============================] - 0s 1ms/step - loss: 91.1068 - val_loss: 56.5816
Epoch 170/300
303/303 [==============================] - 0s 1ms/step - loss: 90.9147 - val_loss: 56.4332
Epoch 171/300
303/303 [==============================] - 0s 1ms/step - loss: 90.7410 - val_loss: 56.3136
Epoch 172/300
303/303 [==============================] - 0s 1ms/step - loss: 90.5381 - val_loss: 56.1210
Epoch 173/300
303/303 [==============================] - 0s 1ms/step - loss: 90.3161 - val_loss: 55.9967
Epoch 174/300
303/303 [==============================] - 0s 1ms/step - loss: 90.1171 - val_loss: 55.8604
Epoch 175/300
303/303 [==============================] 

303/303 [==============================] - 0s 1ms/step - loss: 82.7525 - val_loss: 49.6413
Epoch 245/300
303/303 [==============================] - 0s 1ms/step - loss: 82.7014 - val_loss: 49.5866
Epoch 246/300
303/303 [==============================] - 0s 1ms/step - loss: 82.6279 - val_loss: 49.5633
Epoch 247/300
303/303 [==============================] - 0s 1ms/step - loss: 82.5673 - val_loss: 49.4842
Epoch 248/300
303/303 [==============================] - 0s 1ms/step - loss: 82.4915 - val_loss: 49.4177
Epoch 249/300
303/303 [==============================] - 0s 1ms/step - loss: 82.4331 - val_loss: 49.3623
Epoch 250/300
303/303 [==============================] - 0s 1ms/step - loss: 82.3404 - val_loss: 49.3297
Epoch 251/300
303/303 [==============================] - 0s 1ms/step - loss: 82.3019 - val_loss: 49.2888
Epoch 252/300
303/303 [==============================] - 0s 1ms/step - loss: 82.2583 - val_loss: 49.2240
Epoch 253/300
303/303 [==============================] - 0s 1ms/step 

303/303 [==============================] - 0s 1ms/step - loss: 265.1563 - val_loss: 237.2316
Epoch 23/300
303/303 [==============================] - 0s 1ms/step - loss: 258.5571 - val_loss: 231.7927
Epoch 24/300
303/303 [==============================] - 0s 1ms/step - loss: 252.2722 - val_loss: 226.5998
Epoch 25/300
303/303 [==============================] - 0s 1ms/step - loss: 246.1755 - val_loss: 221.6696
Epoch 26/300
303/303 [==============================] - 0s 1ms/step - loss: 240.4185 - val_loss: 216.9635
Epoch 27/300
303/303 [==============================] - 0s 1ms/step - loss: 235.0818 - val_loss: 212.5063
Epoch 28/300
303/303 [==============================] - 0s 1ms/step - loss: 230.0471 - val_loss: 208.2501
Epoch 29/300
303/303 [==============================] - 0s 1ms/step - loss: 224.9032 - val_loss: 204.2135
Epoch 30/300
303/303 [==============================] - 0s 1ms/step - loss: 220.2923 - val_loss: 200.3871
Epoch 31/300
303/303 [==============================] - 0s 

Epoch 177/300
303/303 [==============================] - 0s 1ms/step - loss: 81.1631 - val_loss: 92.9349
Epoch 178/300
303/303 [==============================] - 0s 1ms/step - loss: 81.0061 - val_loss: 92.8421
Epoch 179/300
303/303 [==============================] - 0s 1ms/step - loss: 80.8792 - val_loss: 92.7126
Epoch 180/300
303/303 [==============================] - 0s 1ms/step - loss: 80.7301 - val_loss: 92.6187
Epoch 181/300
303/303 [==============================] - 0s 1ms/step - loss: 80.5794 - val_loss: 92.4949
Epoch 182/300
303/303 [==============================] - 0s 1ms/step - loss: 80.4497 - val_loss: 92.3437
Epoch 183/300
303/303 [==============================] - 0s 1ms/step - loss: 80.3072 - val_loss: 92.2264
Epoch 184/300
303/303 [==============================] - 0s 1ms/step - loss: 80.1876 - val_loss: 92.0919
Epoch 185/300
303/303 [==============================] - 0s 1ms/step - loss: 80.0336 - val_loss: 91.9469
Epoch 186/300
303/303 [==============================] 

303/303 [==============================] - 0s 1ms/step - loss: 71.5755 - val_loss: 83.9653
Epoch 256/300
303/303 [==============================] - 0s 1ms/step - loss: 71.5389 - val_loss: 83.8448
Epoch 257/300
303/303 [==============================] - 0s 1ms/step - loss: 71.4317 - val_loss: 83.7365
Epoch 258/300
303/303 [==============================] - 0s 1ms/step - loss: 71.3546 - val_loss: 83.5979
Epoch 259/300
303/303 [==============================] - 0s 1ms/step - loss: 71.2753 - val_loss: 83.4143
Epoch 260/300
303/303 [==============================] - 0s 1ms/step - loss: 71.1419 - val_loss: 83.2962
Epoch 261/300
303/303 [==============================] - 0s 1ms/step - loss: 71.0604 - val_loss: 83.1845
Epoch 262/300
303/303 [==============================] - 0s 1ms/step - loss: 70.9451 - val_loss: 83.0341
Epoch 263/300
303/303 [==============================] - 0s 1ms/step - loss: 70.8119 - val_loss: 82.8454
Epoch 264/300
303/303 [==============================] - 0s 1ms/step 

Epoch 33/300
303/303 [==============================] - 0s 1ms/step - loss: 181.8437 - val_loss: 248.2156
Epoch 34/300
303/303 [==============================] - 0s 1ms/step - loss: 178.7975 - val_loss: 244.0107
Epoch 35/300
303/303 [==============================] - 0s 1ms/step - loss: 175.8580 - val_loss: 240.0605
Epoch 36/300
303/303 [==============================] - 0s 1ms/step - loss: 173.1965 - val_loss: 236.2741
Epoch 37/300
303/303 [==============================] - 0s 1ms/step - loss: 170.5627 - val_loss: 232.6703
Epoch 38/300
303/303 [==============================] - 0s 1ms/step - loss: 168.1807 - val_loss: 229.1220
Epoch 39/300
303/303 [==============================] - 0s 1ms/step - loss: 165.7327 - val_loss: 225.6536
Epoch 40/300
303/303 [==============================] - 0s 1ms/step - loss: 163.5462 - val_loss: 222.4934
Epoch 41/300
303/303 [==============================] - 0s 1ms/step - loss: 161.4010 - val_loss: 219.3914
Epoch 42/300
303/303 [========================

303/303 [==============================] - 0s 1ms/step - loss: 78.3821 - val_loss: 113.6476
Epoch 188/300
303/303 [==============================] - 0s 1ms/step - loss: 78.3252 - val_loss: 113.5935
Epoch 189/300
303/303 [==============================] - 0s 1ms/step - loss: 78.2428 - val_loss: 113.5148
Epoch 190/300
303/303 [==============================] - 0s 1ms/step - loss: 78.1807 - val_loss: 113.4598
Epoch 191/300
303/303 [==============================] - 0s 1ms/step - loss: 78.1219 - val_loss: 113.4064
Epoch 192/300
303/303 [==============================] - 0s 1ms/step - loss: 78.0582 - val_loss: 113.3350
Epoch 193/300
303/303 [==============================] - 0s 1ms/step - loss: 77.9833 - val_loss: 113.2761
Epoch 194/300
303/303 [==============================] - 0s 1ms/step - loss: 77.9227 - val_loss: 113.2146
Epoch 195/300
303/303 [==============================] - 0s 1ms/step - loss: 77.8902 - val_loss: 113.1978
Epoch 196/300
303/303 [==============================] - 0s 

In [35]:
for i in range(300):
    print('iteration : ', i+1 ,'loss : ', DA_L[i], 'val_loss : ', DV_L[i])

iteration :  1 loss :  566.4640350341797 val_loss :  555.6436462402344
iteration :  2 loss :  543.4768142700195 val_loss :  531.5094223022461
iteration :  3 loss :  519.0196228027344 val_loss :  507.6222686767578
iteration :  4 loss :  495.5177917480469 val_loss :  484.72010040283203
iteration :  5 loss :  473.01236724853516 val_loss :  462.9796600341797
iteration :  6 loss :  451.7186279296875 val_loss :  442.2658996582031
iteration :  7 loss :  431.29518127441406 val_loss :  422.74063873291016
iteration :  8 loss :  412.1791534423828 val_loss :  404.4472198486328
iteration :  9 loss :  394.47147369384766 val_loss :  387.27622985839844
iteration :  10 loss :  377.96807861328125 val_loss :  371.2413024902344
iteration :  11 loss :  362.27164459228516 val_loss :  356.38587951660156
iteration :  12 loss :  347.79361724853516 val_loss :  342.5811080932617
iteration :  13 loss :  334.28797912597656 val_loss :  329.614990234375
iteration :  14 loss :  321.93543243408203 val_loss :  317.7428

question 8)

Ce code réalise une 4-cross validation, l'intérêt d'entrainer le modèle à partir de 4 différents ensembles d'entrainement et de le tester sur 4 ensembles de validation différents est qu'on obtient un modèle plus fiable, moins biaisé, dont les prédictions test devraient être meilleures, la plupart du temps, que si on avait entraîné puis validé le modèle sur seulement un ensemble à chaque fois.

Globalement on tire les mêmes conclusions de l'analyse qu'à la question 7) , à ceci près que l'erreur de validation devient plus petite à l'époque d'apprentissage 27, et que l'erreur de validation obtenue est plus petite qu'à la question 7), on peut même dire beaucoup plus proche de l'erreur d'entraînement.

En effet, question 7), à l'époque d'apprentissage 300 on avait :

loss: 73.0716 - val_loss: 106.8089 .

Alors qu'ici avec la 4-cross validation on a loss :  67.79207992553711 - val_loss :  74.0708417892456 .

In [29]:
np.size(DA), np.size(Da), np.size(DT), np.size(Dt)

(5252, 404, 1326, 102)

In [30]:
DA.shape, Da.shape, DT.shape, Dt.shape

((404, 13), (404,), (102, 13), (102,))

question 1)

Puisque l'apprentissage se fait selon DA et Da, on en déduit que DA est un tenseur qui représente les données d'entrainement, et que Da représente les valeurs en milliers de dollars qui y sont associées.
DA et Da sont ici des array numpy, alors que DA est une array de 404 vecteurs de 13 éléments Da est une array de 404 valeurs flottantes, ce qui explique que la longueur de DA soit égale à 13 fois la longueur de Da, et qui correspond aux 13 caractéristiques d'une maison définissant un prix en milliers de dollars.
On retrouve la même relation entre la longueur de DT et celle de Dt, puisque DT est le tenseur qui représente les données de test, et que Dt représente les valeurs en milliers de dollars qui y sont associées.

In [7]:
(I,F),(B,E)=boston_housing.load_data()
m=I.mean(axis=0); s=I.std(axis=0); B-=m; B/=s; I-=m; I/=s
A=I.shape[0]//4; G=I[A:]; C=F[A:]
H=I[:A]; D=F[:A]

In [12]:
C[2]

26.6

In [10]:
D[3]

21.1